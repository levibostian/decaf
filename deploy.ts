import { getLastGitHubRelease } from "./lib/github.ts";
import * as git from "./lib/git.ts";

const currentBranch = await git.getCurrentBranchName();

console.log(`current git branch ${currentBranch}`)

// example value for GITHUB_REPOSITORY: "denoland/deno"
const [owner, repo] = Deno.env.get("GITHUB_REPOSITORY")!.split("/"); 

console.log(await getLastGitHubRelease({ owner, repo, branch: currentBranch }));