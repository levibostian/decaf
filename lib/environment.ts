import { exec } from "./exec.ts"
import * as log from "./log.ts"
import { AnyStepName } from "./steps/types/any-step.ts"
import envCi from "env-ci"

export interface Environment {
  getRepository(): { owner: string; repo: string }
  getBuild(): { buildUrl: string; buildId: string; currentBranch: string }
  getSimulatedMergeType(): "merge" | "rebase" | "squash"
  getEventThatTriggeredThisRun(): "push" | "pull_request" | "other"
  isRunningInPullRequest(): { baseBranch: string; targetBranch: string; prNumber: number } | undefined
  getCommandForStep({ stepName }: { stepName: AnyStepName }): string | undefined
  getGitConfigInput(): { name: string; email: string } | undefined
  setOutput({ key, value }: { key: string; value: string }): Promise<void>
  // A catch-all method to get inputs that dont match the other methods.
  getUserConfigurationOptions(): { failOnDeployVerification: boolean; makePullRequestComment: boolean }
}

export class EnvironmentImpl implements Environment {
  private readonly env: Record<string, string>
  private outputFileCache: Record<string, string>

  constructor() {
    // Example of the values in `this.env`:
    // {"isCi":true,"name":"GitHub Actions","service":"github","commit":"7d4aec10df2b2dcbe99643662beca90a24a8a81f","build":"15876053587","isPr":true,"branch":"alpha","prBranch":"refs/pull/68/merge","slug":"levibostian/new-deployment-tool","root":"/home/runner/work/new-deployment-tool/new-deployment-tool","pr":68}
    this.env = envCi()
    this.outputFileCache = {}
  }

  // Note: For pull requests, the return value is pretty useless. Example: "refs/pull/68/merge"
  getNameOfCurrentBranch(): string {
    if (this.env.isPr) return this.env.prBranch
    return this.env.branch
  }

  getBuild(): { buildUrl: string; buildId: string; currentBranch: string } {
    return {
      buildId: this.env.build,
      buildUrl: this.env.buildUrl,
      currentBranch: this.getNameOfCurrentBranch(),
    }
  }

  getRepository(): { owner: string; repo: string } {
    return {
      owner: this.env.slug.split("/")[0],
      repo: this.env.slug.split("/")[1],
    }
  }

  getSimulatedMergeType(): "merge" | "rebase" | "squash" {
    const githubActionInputKey = "simulated_merge_type"

    const simulateMergeType = this.getInput(githubActionInputKey)
    if (!simulateMergeType) {
      return "merge"
    }

    if (simulateMergeType !== "merge" && simulateMergeType !== "rebase" && simulateMergeType !== "squash") {
      log.error(
        `The value for the GitHub Actions input ${githubActionInputKey} is invalid. The value must be either "merge", "rebase", or "squash". The value provided was: ${simulateMergeType}`,
      )

      throw new Error()
    }

    return simulateMergeType
  }

  getEventThatTriggeredThisRun(): "push" | "pull_request" | "other" {
    if (this.env.isPr) return "pull_request"
    if (this.env.branch) return "push"

    return "other"
  }

  isRunningInPullRequest(): { baseBranch: string; targetBranch: string; prNumber: number } | undefined {
    if (!this.env.isPr) {
      return undefined
    }

    return {
      baseBranch: this.env.prBranch,
      targetBranch: this.env.branch,
      prNumber: Number(this.env.pr), // module provides this as a string, but we want it as a number.
    }
  }

  async setOutput({ key, value }: { key: string; value: string }): Promise<void> {
    await this.setOutputFile(key, value)

    if (this.env.service == "github") {
      await exec.run({
        command: `echo "${key}=${value}" >> "$GITHUB_OUTPUT"`,
        input: undefined,
      })
    }

    // Open to adding other CI services in the future, if they have a simple system for setting outputs.
  }

  getUserConfigurationOptions(): { failOnDeployVerification: boolean; makePullRequestComment: boolean } {
    return {
      failOnDeployVerification: this.getInput("fail_on_deploy_verification") === "true",
      makePullRequestComment: this.getInput("make_pull_request_comment") === "true",
    }
  }

  getGitConfigInput(): { name: string; email: string } | undefined {
    const gitConfigInput = this.getInput("git_config")
    if (!gitConfigInput || gitConfigInput.trim() === "") {
      return undefined
    }

    // Expect format: "Name <email>"
    const match = gitConfigInput.match(/^(.*)\s+<([^>]+)>$/)
    if (!match) {
      log.error(
        `The git_config input must be in the format "name <email>". The value provided was: ${gitConfigInput}`,
      )
      throw new Error()
    }

    const name = match[1].trim()
    const email = match[2].trim()

    return { name, email }
  }

  getCommandForStep({ stepName }: { stepName: string }): string | undefined {
    const command = this.getInput(stepName)
    if (!command) return undefined
    return command
  }

  private getInput(key: string): string {
    const val: string = Deno.env.get(`INPUT_${key.replace(/ /g, "_").toUpperCase()}`) || ""
    if (!val) {
      throw new Error(`The GitHub Actions input "${key}" is not set.`)
    }

    return val.trim()
  }

  // Writes output to a file as a way to support other CI services that are not github actions.
  private async setOutputFile(key: string, value: string): Promise<void> {
    const nameOfOutputFile = Deno.env.get("INPUT_OUTPUT_FILE")
    if (!nameOfOutputFile || nameOfOutputFile.trim() === "") return

    this.outputFileCache[key] = value

    await Deno.writeFile(
      nameOfOutputFile,
      new TextEncoder().encode(JSON.stringify(this.outputFileCache, null, 2)),
    )
  }
}
