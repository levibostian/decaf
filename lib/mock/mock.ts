import { stub, spy, GetParametersFromProp, GetReturnFromProp, Stub } from "jsr:@std/testing@1/mock";

/**
 * Example: 
 * ```
 * interface GitHubActions {
 *  isRunningInPullRequest(): Promise<boolean>
 * }
 * 
 * const githubActions: GitHubActions = mock()
 * when(githubActions, "isRunningInPullRequest", async () => true)
 * 
 * // optionally, get information about the mock: 
 * const isRunningInPullRequestMock = when(...)
 * assertEquals(isRunningInPullRequestMock.calls.length, 1)
 * ```
 */

export function when<T, Prop extends keyof T>(
  mockObject: T, 
  functionName: Prop, 
  thenCall: (this: T, ...args: GetParametersFromProp<T, Prop>) => GetReturnFromProp<T, Prop>): Stub<T, GetParametersFromProp<T, Prop>, GetReturnFromProp<T, Prop>> {
    return stub(mockObject, functionName, thenCall)
}

export function mock<T>(): T {
  const mockObject = {} as T;
  
  // for convenience, stub all functions with an empty implementation. So if a mock function doesn't need any logic, it can be used without needing to call `when` on it.
  for (const key in mockObject) {
    mockObject[key] = stub(mockObject, key) as unknown as T[Extract<keyof T, string>];
  }
  
  return mockObject;
}
