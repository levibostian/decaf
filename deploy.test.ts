import { assertEquals, assertRejects } from "@std/assert"
import { afterEach, describe, it } from "@std/testing/bdd"
import { restore, stub } from "@std/testing/mock"
import { assertSnapshot } from "@std/testing/snapshot"
import { run } from "./deploy.ts"
import { GetCommitsSinceLatestReleaseStep } from "./lib/steps/get-commits-since-latest-release.ts"
import { getLogMock } from "./lib/log.test.ts"
import { Environment } from "./lib/environment.ts"
import { PrepareTestModeEnvStep } from "./lib/steps/prepare-testmode-env.ts"
import { mock, when } from "./lib/mock/mock.ts"
import { StepRunner } from "./lib/step-runner.ts"
import { GetLatestReleaseStepOutputFake } from "./lib/steps/types/output.test.ts"
import { GetLatestReleaseStepOutput } from "./lib/steps/types/output.ts"
import { ConvenienceStep } from "./lib/steps/convenience.ts"
import { GitCommit } from "./lib/types/git.ts"
import { GitCommitFake } from "./lib/types/git.test.ts"

describe("run the tool in different scenarios", () => {
  afterEach(() => {
    restore()
  })

  it("given no commits created since last deployment, expect to not run a new deployment", async () => {
    const { determineNextReleaseVersionStepMock } = await setupTestEnvironmentAndRun({
      commitsSinceLatestRelease: [],
    })

    // Exit early, before running the next step after getting list of commits since last deployment
    assertEquals(determineNextReleaseVersionStepMock.calls.length, 0)
  })

  it("given no commits trigger a release, expect to not run a new deployment", async () => {
    const { deployStepMock } = await setupTestEnvironmentAndRun({
      commitsSinceLatestRelease: [new GitCommitFake()],
      nextReleaseVersion: undefined,
    })

    // Exit early, before running the next step after getting list of commits since last deployment
    assertEquals(deployStepMock.calls.length, 0)
  })
})

describe("test github actions output", () => {
  it("should set new release version output when a new release is created", async () => {
    const { setOutputMock } = await setupTestEnvironmentAndRun({
      commitsSinceLatestRelease: [
        new GitCommitFake({
          message: "feat: trigger a release",
          sha: "trigger-release",
        }),
      ],
      nextReleaseVersion: "1.0.0",
    })

    assertEquals(
      setOutputMock.calls[1].args[0],
      { key: "new_release_version", value: "1.0.0" },
    )
  })
  it("should not set new release version output when no new release is created", async () => {
    const { setOutputMock } = await setupTestEnvironmentAndRun({
      commitsSinceLatestRelease: [],
      nextReleaseVersion: "1.0.0",
    })

    assertEquals(setOutputMock.calls.filter((call) => call.args[0].key === "new_release_version").length, 0)
  })
  it("should set new pre-release version output when a new pre-release is created", async () => {
    const { setOutputMock } = await setupTestEnvironmentAndRun({
      commitsSinceLatestRelease: [
        new GitCommitFake({
          message: "feat: trigger a release",
          sha: "trigger-release",
        }),
      ],
      nextReleaseVersion: "1.0.0-beta.1",
    })

    assertEquals(
      setOutputMock.calls[1].args[0],
      { key: "new_release_version", value: "1.0.0-beta.1" },
    )
  })
  it("should set test mode output when running in test mode", async () => {
    const { setOutputMock } = await setupTestEnvironmentAndRun({
      commitsSinceLatestRelease: [
        new GitCommitFake({
          message: "feat: trigger a release",
          sha: "trigger-release",
        }),
      ],
      nextReleaseVersion: "1.0.0",
      githubActionEventThatTriggeredTool: "pull_request",
      commitsCreatedBySimulatedMerge: [new GitCommitFake()],
    })

    assertEquals(
      setOutputMock.calls[0].args[0],
      { key: "test_mode_on", value: "true" },
    )
  })
  it("should add commits created during simulated merges to list of commits to analyze", async () => {
    const givenLatestCommitOnBranch = new GitCommitFake({
      message: "feat: trigger a release",
      sha: "trigger-release",
    })

    const givenCommitsCreatedBySimulatedMerge = [
      new GitCommitFake({
        message: "Merge commit created during simulated merge",
        sha: "merge-commit-created-during-simulated-merge",
      }),
      new GitCommitFake({
        message: "feat: commit created during simulated merge",
        sha: "commit-created-during-simulated-merge",
      }),
    ]

    const expectedCommitsAnalyzed = [
      givenCommitsCreatedBySimulatedMerge[0], // newest commit first. like `git log`
      givenCommitsCreatedBySimulatedMerge[1],
      givenLatestCommitOnBranch,
    ]

    const { determineNextReleaseVersionStepMock } = await setupTestEnvironmentAndRun({
      pullRequestTargetBranchName: "main",
      currentBranchName: "sweet-feature",
      githubActionEventThatTriggeredTool: "pull_request",
      commitsSinceLatestRelease: [...givenCommitsCreatedBySimulatedMerge, givenLatestCommitOnBranch],
      nextReleaseVersion: "1.0.0",
      commitsCreatedBySimulatedMerge: givenCommitsCreatedBySimulatedMerge,
    })

    assertEquals(determineNextReleaseVersionStepMock.calls[0].args[0].gitCommitsSinceLastRelease, expectedCommitsAnalyzed)
  })

  // Test a bug found where you are in test mode > simulate merge > new commits are made > tool says "zero commits created" > exits early.
  it("should not exit early if parent branch has no commits, but we make new commits during simulated merge", async () => {
    const givenCommitsCreatedBySimulatedMerge = [
      new GitCommitFake({
        message: "Merge commit created during simulated merge",
        sha: "merge-commit-created-during-simulated-merge",
      }),
      new GitCommitFake({
        message: "feat: commit created during simulated merge",
        sha: "commit-created-during-simulated-merge",
      }),
    ]

    const { setOutputMock } = await setupTestEnvironmentAndRun({
      pullRequestTargetBranchName: "main",
      currentBranchName: "sweet-feature",
      githubActionEventThatTriggeredTool: "pull_request",
      commitsSinceLatestRelease: givenCommitsCreatedBySimulatedMerge, // the only commits we have are the ones created during simulated merge
      nextReleaseVersion: "1.0.0",
      commitsCreatedBySimulatedMerge: givenCommitsCreatedBySimulatedMerge,
    })

    const didExitEarly = setOutputMock.calls.filter((call) => call.args[0].key === "new_release_version").length === 0

    assertEquals(didExitEarly, false)
  })
})

describe("user facing logs", () => {
  it("given no commits will trigger a release, expect logs to easily communicate that to the user", async (t) => {
    const { logMock } = await setupTestEnvironmentAndRun({
      commitsSinceLatestRelease: [new GitCommitFake()],
      nextReleaseVersion: undefined,
    })

    await assertSnapshot(t, logMock.getLogs({ includeDebugLogs: false }))
  })

  it("given no commits created since last deployment, expect logs to easily communicate that to the user", async (t) => {
    const { logMock } = await setupTestEnvironmentAndRun({
      commitsSinceLatestRelease: [],
    })

    await assertSnapshot(t, logMock.getLogs({ includeDebugLogs: false }))
  })

  it("given no release has ever been made, expect logs to easily communicate that to the user", async (t) => {
    const givenLatestCommitOnBranch = new GitCommitFake({
      message: "feat: trigger a release",
      sha: "trigger-release",
    })

    const { logMock } = await setupTestEnvironmentAndRun({
      latestRelease: null,
      commitsSinceLatestRelease: [givenLatestCommitOnBranch],
      nextReleaseVersion: "1.0.0",
    })

    await assertSnapshot(t, logMock.getLogs({ includeDebugLogs: false }))
  })

  it("given running in test mode, given commits that trigger a release, expect logs to easily communicate that to the user", async (t) => {
    const givenBaseBranch = "sweet-feature"
    const givenTargetBranch = "main"

    const givenLatestCommitOnBranch = new GitCommitFake({
      message: "feat: trigger a release",
      sha: "trigger-release",
    })

    const { logMock } = await setupTestEnvironmentAndRun({
      pullRequestTargetBranchName: givenTargetBranch,
      currentBranchName: givenBaseBranch,
      githubActionEventThatTriggeredTool: "pull_request",
      commitsSinceLatestRelease: [givenLatestCommitOnBranch],
      nextReleaseVersion: "1.0.0",
    })

    await assertSnapshot(t, logMock.getLogs({ includeDebugLogs: false }))
  })
})

describe("test the event that triggered running the tool", () => {
  it("should exit early if the tool is triggered by an unsupported event", async () => {
    const { runGetLatestOnCurrentBranchReleaseStepMock, prepareEnvironmentForTestModeMock } = await setupTestEnvironmentAndRun({
      githubActionEventThatTriggeredTool: "other",
    })

    assertEquals(runGetLatestOnCurrentBranchReleaseStepMock.calls.length, 0)
    assertEquals(prepareEnvironmentForTestModeMock.calls.length, 0)
  })
  it("should run a deployment if triggered from a push event, expect not to run simulated merge", async () => {
    const { setOutputMock, prepareEnvironmentForTestModeMock } = await setupTestEnvironmentAndRun({
      githubActionEventThatTriggeredTool: "push",
      nextReleaseVersion: "1.0.0",
      commitsSinceLatestRelease: [
        new GitCommitFake({
          message: "feat: trigger a release",
          sha: "trigger-release",
        }),
      ],
    })

    assertEquals(setOutputMock.calls[1].args[0].value, "1.0.0")
    assertEquals(setOutputMock.calls[0].args[0], { key: "test_mode_on", value: "false" })
    assertEquals(prepareEnvironmentForTestModeMock.calls.length, 0)
  })
  it("should run a deployment in test mode if triggered from a pull_request event, expect to run merge simulation", async () => {
    const givenBaseBranch = "sweet-feature"
    const givenTargetBranch = "main"

    const { setOutputMock, prepareEnvironmentForTestModeMock, runGetLatestOnCurrentBranchReleaseStepMock } = await setupTestEnvironmentAndRun({
      pullRequestTargetBranchName: givenTargetBranch,
      currentBranchName: givenBaseBranch,
      githubActionEventThatTriggeredTool: "pull_request",
      nextReleaseVersion: "1.0.0",
      commitsSinceLatestRelease: [
        new GitCommitFake({
          message: "feat: trigger a release",
          sha: "trigger-release",
        }),
      ],
    })

    assertEquals(
      setOutputMock.calls[1].args[0],
      { key: "new_release_version", value: "1.0.0" },
    )
    assertEquals(setOutputMock.calls[0].args[0], { key: "test_mode_on", value: "true" })
    assertEquals(prepareEnvironmentForTestModeMock.calls.length, 1)

    // We expect that after simulated merge, the current branch is now the target branch.
    assertEquals(runGetLatestOnCurrentBranchReleaseStepMock.calls[0].args[0].gitCurrentBranch, givenTargetBranch)
  })
})

describe("deployment verification after deploy", () => {
  it("should verify the latest release matches the deployed version after deployment", async () => {
    const deployedVersion = "2.0.0"
    const { setOutputMock } = await setupTestEnvironmentAndRun({
      latestRelease: { versionName: "1.0.0", commitSha: "sha1" },
      latestReleaseAfterDeploy: { versionName: deployedVersion, commitSha: "sha2" },
      nextReleaseVersion: deployedVersion,
      commitsSinceLatestRelease: [new GitCommitFake()],
    })

    assertEquals(
      setOutputMock.calls[1].args[0],
      { key: "new_release_version", value: deployedVersion },
    )
  })

  it("should throw if the latest release does not match the deployed version after deployment", async () => {
    const deployedVersion = "2.0.0"
    const wrongLatestRelease = "1.5.0"

    await assertRejects(async () => {
      await setupTestEnvironmentAndRun({
        latestRelease: { versionName: wrongLatestRelease, commitSha: "sha2" },
        latestReleaseAfterDeploy: { versionName: wrongLatestRelease, commitSha: "sha2" },
        nextReleaseVersion: deployedVersion,
        commitsSinceLatestRelease: [new GitCommitFake()],
      })
    })
  })
})

describe("input data passed to steps", () => {
  it("should pass correct input data to determineNextReleaseVersionStep", async () => {
    const givenCommitsCreatedBySimulatedMerge = new GitCommitFake({
      message: "Merge commit",
    })
    const givenCommitsOnBranch = [
      givenCommitsCreatedBySimulatedMerge,
      new GitCommitFake({ message: "feat: commit on branch" }),
      new GitCommitFake({ message: "feat: another commit on branch" }),
    ]

    const { determineNextReleaseVersionStepMock } = await setupTestEnvironmentAndRun({
      commitsSinceLatestRelease: givenCommitsOnBranch,
      commitsCreatedBySimulatedMerge: [givenCommitsCreatedBySimulatedMerge],
      nextReleaseVersion: "1.0.0",
    })

    assertEquals(
      determineNextReleaseVersionStepMock.calls[0].args[0].gitCommitsSinceLastRelease,
      givenCommitsOnBranch,
    )
  })
})

const setupTestEnvironmentAndRun = async ({
  latestRelease,
  latestReleaseAfterDeploy,
  commitsSinceLatestRelease,
  nextReleaseVersion,
  githubActionEventThatTriggeredTool,
  pullRequestTargetBranchName,
  currentBranchName,
  commitsCreatedBySimulatedMerge,
  gitCommitsCurrentBranch,
}: {
  latestRelease?: GetLatestReleaseStepOutput | null
  latestReleaseAfterDeploy?: GetLatestReleaseStepOutput | null
  commitsSinceLatestRelease?: GitCommit[]
  nextReleaseVersion?: string
  githubActionEventThatTriggeredTool?: "push" | "pull_request" | "other"
  pullRequestTargetBranchName?: string
  currentBranchName?: string
  commitsCreatedBySimulatedMerge?: GitCommit[]
  gitCommitsCurrentBranch?: GitCommit[]
}) => {
  // Set some defaults.
  const pullRequestTargetBranch = pullRequestTargetBranchName || "main" // assume we are running a pull_request event that merges into main
  const currentBranch = currentBranchName || "main" // assume we are running a push event

  // default to push event, since we want to test the actual deployment process and not test mode by default.
  Deno.env.set("GITHUB_REF", `refs/heads/${currentBranch}`)
  Deno.env.set("GITHUB_REPOSITORY", "levibostian/decaf")

  const convenienceStep = mock<ConvenienceStep>()
  when(convenienceStep, "runConvenienceCommands", async () => {
    return { gitCommitsAllLocalBranches: { "latest": [] }, gitCommitsCurrentBranch: gitCommitsCurrentBranch || [] }
  })

  const stepRunner = {} as StepRunner
  const runGetLatestOnCurrentBranchReleaseStepMock = stub(
    stepRunner,
    "runGetLatestOnCurrentBranchReleaseStep",
    async () => {
      // if we ran a deployment, return a different value since latest release might have changed.
      if (deployStepMock.calls.length > 0) {
        if (latestReleaseAfterDeploy) return latestReleaseAfterDeploy
        if (nextReleaseVersion) return { versionName: nextReleaseVersion, commitSha: "deploy-sha" }
      }

      if (latestRelease === null) return null
      if (latestRelease === undefined) return GetLatestReleaseStepOutputFake
      return latestRelease
    },
  )
  const determineNextReleaseVersionStepMock = stub(
    stepRunner,
    "determineNextReleaseVersionStep",
    async () => {
      if (nextReleaseVersion) return { version: nextReleaseVersion }
      return null
    },
  )

  const getCommitsSinceLatestReleaseStep = {} as GetCommitsSinceLatestReleaseStep
  const getCommitsSinceLatestReleaseStepMock = stub(
    getCommitsSinceLatestReleaseStep,
    "getAllCommitsSinceGivenCommit",
    async () => {
      return commitsSinceLatestRelease || []
    },
  )

  const deployStepMock = stub(stepRunner, "runDeployStep", async () => {
    return
  })

  const logMock = getLogMock()

  const environment = {} as Environment
  const setOutputMock = stub(
    environment,
    "setOutput",
    async () => {
      return
    },
  )
  stub(environment, "getEventThatTriggeredThisRun", () => {
    return githubActionEventThatTriggeredTool || "push"
  })
  stub(environment, "getUserConfigurationOptions", () => {
    return {
      failOnDeployVerification: true,
      makePullRequestComment: true,
    }
  })
  const isRunningInPullRequest = githubActionEventThatTriggeredTool === "pull_request"

  const environmentIsRunningInPullRequestMock = stub(environment, "isRunningInPullRequest", () => {
    return isRunningInPullRequest
      ? {
        baseBranch: currentBranch,
        targetBranch: pullRequestTargetBranch,
        prNumber: 1,
      }
      : undefined
  })
  stub(environment, "getSimulatedMergeType", (): "merge" | "rebase" | "squash" => {
    return "merge"
  })

  stub(environment, "getBuild", () => {
    return {
      currentBranch,
      buildUrl: "https://example.com/build/123",
      buildId: "123",
    }
  })
  stub(environment, "getRepository", () => {
    return { owner: "levibostian", repo: "decaf" }
  })

  // for simplicity, we return default values for these.
  stub(environment, "getBranchFilters", () => {
    return []
  })
  stub(environment, "getCommitLimit", () => {
    return 500
  })

  const prepareEnvironmentForTestMode = mock<PrepareTestModeEnvStep>()
  const prepareEnvironmentForTestModeMock = when(prepareEnvironmentForTestMode, "prepareEnvironmentForTestMode", async () => {
    if (!isRunningInPullRequest) return undefined

    return { currentGitBranch: pullRequestTargetBranch, commitsCreatedDuringSimulatedMerges: commitsCreatedBySimulatedMerge || [] }
  })

  await run({
    convenienceStep,
    stepRunner,
    prepareEnvironmentForTestMode,
    getCommitsSinceLatestReleaseStep,
    log: logMock,
    environment,
  })

  return {
    runGetLatestOnCurrentBranchReleaseStepMock,
    getCommitsSinceLatestReleaseStepMock,
    determineNextReleaseVersionStepMock,
    deployStepMock,
    logMock,
    setOutputMock,
    environmentIsRunningInPullRequestMock,
    prepareEnvironmentForTestModeMock,
  }
}
