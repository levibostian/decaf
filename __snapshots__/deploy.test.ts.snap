export const snapshot = {};

snapshot[`user facing logs > given no commits will trigger a release, expect logs to easily communicate that to the user 1`] = `
[
  "👋 Hello! I am a tool called new-deployment-tool. I help you deploy your projects.",
  "To learn how the deployment process of your project works, I suggest reading all of the logs that I print to you below.",
  "If you have more questions after reading the logs, you can optionally view the documentation to learn more about the tool: https://github.com/levibostian/new-deployment-tool/",
  "Ok, let's get started with the deployment!",
  "--------------------------------",
  "👀 I see that the git branch main is checked out. We will begin the deployment process from the latest commit of this branch.",
  "🔍 First, I need to get the latest release that was created on the git branch main. I'll look for it now...",
  "I have been told that the latest release on the git branch main is: v1.0.0",
  "📜 Next, I need to know all of the changes (git commits) that have been done on git branch main since the latest release: v1.0.0, commit: abc123. I'll look for them now...",
  "I found 1 git commits created since the latest release of v1.0.0.",
  "📊 Now I need to know (1) if any of these new commits need to be deployed and (2) if they should, what should the new version be. To determine this, I will analyze each git commit one-by-one...",
  "After analyzing all of the git commits, none of the commits need to be deployed. Therefore, the deployment process stops here with no new release to be made. Bye-bye 👋!",
]
`;

snapshot[`user facing logs > given no commits created since last deployment, expect logs to easily communicate that to the user 1`] = `
[
  "👋 Hello! I am a tool called new-deployment-tool. I help you deploy your projects.",
  "To learn how the deployment process of your project works, I suggest reading all of the logs that I print to you below.",
  "If you have more questions after reading the logs, you can optionally view the documentation to learn more about the tool: https://github.com/levibostian/new-deployment-tool/",
  "Ok, let's get started with the deployment!",
  "--------------------------------",
  "👀 I see that the git branch main is checked out. We will begin the deployment process from the latest commit of this branch.",
  "🔍 First, I need to get the latest release that was created on the git branch main. I'll look for it now...",
  "I have been told that the latest release on the git branch main is: v1.0.0",
  "📜 Next, I need to know all of the changes (git commits) that have been done on git branch main since the latest release: v1.0.0, commit: abc123. I'll look for them now...",
  "Looks like zero commits have been created since the latest release. This means there is no new code created and therefore, the deployment process stops here. Bye-bye 👋!",
]
`;

snapshot[`user facing logs > given new commit created during deployment, expect logs to easily communicate that to the user 1`] = `
[
  "👋 Hello! I am a tool called new-deployment-tool. I help you deploy your projects.",
  "To learn how the deployment process of your project works, I suggest reading all of the logs that I print to you below.",
  "If you have more questions after reading the logs, you can optionally view the documentation to learn more about the tool: https://github.com/levibostian/new-deployment-tool/",
  "Ok, let's get started with the deployment!",
  "--------------------------------",
  "👀 I see that the git branch main is checked out. We will begin the deployment process from the latest commit of this branch.",
  "🔍 First, I need to get the latest release that was created on the git branch main. I'll look for it now...",
  "I have been told that the latest release on the git branch main is: v1.0.0",
  "📜 Next, I need to know all of the changes (git commits) that have been done on git branch main since the latest release: v1.0.0, commit: abc123. I'll look for them now...",
  "I found 1 git commits created since the latest release of v1.0.0.",
  "📊 Now I need to know (1) if any of these new commits need to be deployed and (2) if they should, what should the new version be. To determine this, I will analyze each git commit one-by-one...",
  "After analyzing all of the git commits, I have determined the next release version will be: 1.0.0",
  "🚢 It's time to ship 1.0.0! I will now run all of the deployment commands provided in your project's configuration file...",
  "✏️ The code has been shipped. The final piece of the deployment process is creating a new release on GitHub for the new version, 1.0.0. Creating that now...",
  "New release has been created on GitHub. View the release: https://github.com/levibostian/new-deployment-tool/releases/1.0.0",
  "🎉 Congratulations! The deployment process has completed. Bye-bye 👋!",
]
`;

snapshot[`user facing logs > given no new commits created during deployment, expect logs to easily communicate that to the user 1`] = `
[
  "👋 Hello! I am a tool called new-deployment-tool. I help you deploy your projects.",
  "To learn how the deployment process of your project works, I suggest reading all of the logs that I print to you below.",
  "If you have more questions after reading the logs, you can optionally view the documentation to learn more about the tool: https://github.com/levibostian/new-deployment-tool/",
  "Ok, let's get started with the deployment!",
  "--------------------------------",
  "👀 I see that the git branch main is checked out. We will begin the deployment process from the latest commit of this branch.",
  "🔍 First, I need to get the latest release that was created on the git branch main. I'll look for it now...",
  "I have been told that the latest release on the git branch main is: v1.0.0",
  "📜 Next, I need to know all of the changes (git commits) that have been done on git branch main since the latest release: v1.0.0, commit: abc123. I'll look for them now...",
  "I found 1 git commits created since the latest release of v1.0.0.",
  "📊 Now I need to know (1) if any of these new commits need to be deployed and (2) if they should, what should the new version be. To determine this, I will analyze each git commit one-by-one...",
  "After analyzing all of the git commits, I have determined the next release version will be: 1.0.0",
  "🚢 It's time to ship 1.0.0! I will now run all of the deployment commands provided in your project's configuration file...",
  "✏️ The code has been shipped. The final piece of the deployment process is creating a new release on GitHub for the new version, 1.0.0. Creating that now...",
  "New release has been created on GitHub. View the release: https://github.com/levibostian/new-deployment-tool/releases/1.0.0",
  "🎉 Congratulations! The deployment process has completed. Bye-bye 👋!",
]
`;

snapshot[`user facing logs > given no release has ever been made, expect logs to easily communicate that to the user 1`] = `
[
  "👋 Hello! I am a tool called new-deployment-tool. I help you deploy your projects.",
  "To learn how the deployment process of your project works, I suggest reading all of the logs that I print to you below.",
  "If you have more questions after reading the logs, you can optionally view the documentation to learn more about the tool: https://github.com/levibostian/new-deployment-tool/",
  "Ok, let's get started with the deployment!",
  "--------------------------------",
  "👀 I see that the git branch main is checked out. We will begin the deployment process from the latest commit of this branch.",
  "🔍 First, I need to get the latest release that was created on the git branch main. I'll look for it now...",
  "I have been told that the git branch, main, has never been released before. This will be the first release. Exciting!",
  "📜 Next, I need to know all of the changes (git commits) that have been done on git branch main since the latest release: undefined, commit: undefined. I'll look for them now...",
  "I found 1 git commits created since the git branch main was created.",
  "📊 Now I need to know (1) if any of these new commits need to be deployed and (2) if they should, what should the new version be. To determine this, I will analyze each git commit one-by-one...",
  "After analyzing all of the git commits, I have determined the next release version will be: 1.0.0",
  "🚢 It's time to ship 1.0.0! I will now run all of the deployment commands provided in your project's configuration file...",
  "✏️ The code has been shipped. The final piece of the deployment process is creating a new release on GitHub for the new version, 1.0.0. Creating that now...",
  "New release has been created on GitHub. View the release: https://github.com/levibostian/new-deployment-tool/releases/1.0.0",
  "🎉 Congratulations! The deployment process has completed. Bye-bye 👋!",
]
`;

snapshot[`user facing logs > given running in test mode, given commits that trigger a release, expect logs to easily communicate that to the user 1`] = `
[
  "👋 Hello! I am a tool called new-deployment-tool. I help you deploy your projects.",
  "To learn how the deployment process of your project works, I suggest reading all of the logs that I print to you below.",
  "If you have more questions after reading the logs, you can optionally view the documentation to learn more about the tool: https://github.com/levibostian/new-deployment-tool/",
  "Ok, let's get started with the deployment!",
  "--------------------------------",
  "🧪 I see that I got triggered to run from a pull request event. In pull requests, I run in test mode which means that I will run the deployment process but I will not actually deploy anything.",
  "🧪 In test mode, I also simulate merging the current pull request and all parent pull requests (Note, I don't actually merge any pull requests). Simulating now...",
  "🧪 Simulated merges complete. You will notice that for the remainder of the deployment process, the current branch will be main instead of the pull request branch sweet-feature.",
  "👀 I see that the git branch main is checked out. We will begin the deployment process from the latest commit of this branch.",
  "🔍 First, I need to get the latest release that was created on the git branch main. I'll look for it now...",
  "I have been told that the latest release on the git branch main is: v1.0.0",
  "📜 Next, I need to know all of the changes (git commits) that have been done on git branch main since the latest release: v1.0.0, commit: abc123. I'll look for them now...",
  "I found 1 git commits created since the latest release of v1.0.0.",
  "📊 Now I need to know (1) if any of these new commits need to be deployed and (2) if they should, what should the new version be. To determine this, I will analyze each git commit one-by-one...",
  "After analyzing all of the git commits, I have determined the next release version will be: 1.0.0",
  "🚢 It's time to ship 1.0.0! I will now run all of the deployment commands provided in your project's configuration file...",
  "✏️ The code has been shipped. The final piece of the deployment process is creating a new release on GitHub for the new version, 1.0.0. Creating that now...(Test mode is enabled so I will not create the new release)",
  "🎉 Congratulations! The deployment process has completed. Bye-bye 👋!",
]
`;
